compiler/plc_lowering/src/inheritance.rs

```diff
@@ -0,0 +1,1863 @@
+        self.ctx.provider()
+    }
+
+    pub fn visit_unit(&mut self, unit: &mut CompilationUnit) {
+        // XXX: not sure if we need to visit global vars and user-data-types when walking the compilation unit here
+        // for pou in unit.units.iter_mut() {
+        //     pou.walk(self);
+        // }
+        // for implementation in unit.implementations.iter_mut() {
+        //     implementation.walk(self);
+        // }
+        self.visit_compilation_unit(unit);
+    }
+
+    fn walk_with_context<T: WalkerMut>(&mut self, t: &mut T, ctx: Context) {
+        let old_ctx = std::mem::replace(&mut self.ctx, ctx);
+        t.walk(self);
+        self.ctx = old_ctx;
+    }
+
+    // Updates the base to reflect the inheritance chain of the node and returns it
+    fn update_inheritance_chain(
+        &self,
+        node: &AstNode,
+        base: Option<Box<AstNode>>,
+        base_type: &str,
+    ) -> Option<Box<AstNode>> {
+        if self.index.is_none() || self.annotations.is_none() {
+            return base;
+        }
+        // disect the qualified name
+        // a.b.c -> am i a direct member of b?
+        //      b => am I a direct member of a?
+        //      a => am I a direct member of local container?
+
+        let annotations = self.annotations.as_ref().expect("Annotations exist");
+        let Some(qualified_name) = annotations.get_qualified_name(node) else {
+            // this is likely an index into an array/pointer dereference. just return as is
+            return base;
+        };
+
+        let segment = qualified_name.split('.').next().expect("Must have a name");
+
+        if base_type == segment {
+            return base;
+        }
+
+        let index = self.index.as_ref().expect("Index exists");
+        let inheritance_chain = index.find_ancestors(base_type, segment);
+        if inheritance_chain.len() <= 1 {
+            return base;
+        }
+
+        // add a `__BASE` qualifier for each element in the inheritance chain, exluding `self`
+        inheritance_chain.iter().skip(1).fold(base, |base, _| {
+            Some(Box::new(AstFactory::create_member_reference(
+                AstFactory::create_identifier(
+                    "__BASE",
+                    SourceLocation::internal(),
+                    self.provider().next_id(),
+                ),
+                base.map(|it: Box<AstNode>| *it),
+                self.provider().next_id(),
+            )))
+        })
+    }
+}
+
+impl AstVisitorMut for InheritanceLowerer {
+    fn visit_pou(&mut self, pou: &mut Pou) {
+        if self.index.is_some() {
+            return self.walk_with_context(pou, self.ctx.with_pou(&pou.name));
+        }
+        if !matches!(pou.kind, PouType::FunctionBlock | PouType::Class) {
+            return;
+        }
+
+        let Some(base_name) = pou.super_class.as_ref() else {
+            return;
+        };
+
+        let base_var = Variable {
+            name: "__BASE".to_string(),
```

Comment:
is base the correct name? should we maybe go with __super or something? 
We will eventually need a super pointer